aesd-char-driver/main.c

```diff
@@ -37,14 +40,112 @@ static ssize_t aesd_write(struct file *filp, const char __user *buf,
+  PDEBUG("ioctl with command %u", cmd);
+
+  switch (cmd) {
+  case AESDCHAR_IOCSEEKTO:
+    struct aesd_seekto seekto;
+    if (copy_from_user(&seekto, (const void __user *)arg, sizeof(seekto)) !=
+        0) {
+      return -EFAULT;
+    }
 
-struct file_operations aesd_fops = {
-    .owner = THIS_MODULE,
-    .read = aesd_read,
-    .write = aesd_write,
-    .open = aesd_open,
-    .release = aesd_release,
-};
+    return aesd_adjust_file_offset(filp, seekto.write_cmd,
+                                   seekto.write_cmd_offset);
+
+  default:
+    PDEBUG("unhandled ioctl command: %u", cmd);
+    return -EINVAL;
+  }
+}
+
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd,
+                                    unsigned int write_cmd_offset) {
+  PDEBUG("aesd_adujust_file_offset with index %u and offset %u", write_cmd,
+         write_cmd_offset);
+  struct aesd_dev *dev_ptr = (struct aesd_dev *)(filp->private_data);
+  if (mutex_lock_interruptible(&dev_ptr->device_mutex)) {
+    return -ERESTARTSYS;
+  }
+
+  const struct aesd_buffer_entry *buffer_entry =
+      aesd_circular_buffer_get_entry_by_index(write_cmd,
+                                              &(dev_ptr->circular_buffer));
+  if (NULL == buffer_entry || buffer_entry->size <= write_cmd_offset) {
+    mutex_unlock(&dev_ptr->device_mutex);
+    PDEBUG("No data a requested index (%u) and offset (%u)", write_cmd,
+           write_cmd_offset);
+    return -EINVAL;
+  }
+
+  const ssize_t entry_pos = aesd_circular_buffer_get_entry_offset(
+      buffer_entry, &(dev_ptr->circular_buffer));
+  if (entry_pos < 0) {
+    mutex_unlock(&dev_ptr->device_mutex);
+    return -EINVAL;
+  }
+  mutex_unlock(&dev_ptr->device_mutex);
+
+  const size_t file_offset = entry_pos + write_cmd_offset;
+  filp->f_pos += file_offset;
+  return 0;
+}
+
+static loff_t aesd_llseek(struct file *filp, loff_t offset, int whence) {
+  PDEBUG("llseek with offset %lld", offset);
+
+  struct aesd_dev *dev_ptr = (struct aesd_dev *)(filp->private_data);
+  mutex_lock(&dev_ptr->device_mutex);
+
+  ssize_t circular_buffer_size =
+      aesd_circular_buffer_get_size(&(dev_ptr->circular_buffer));
+  if (circular_buffer_size < 0) {
+    PDEBUG("aesd_circular_buffer_get_size returned an error");
+    mutex_unlock(&dev_ptr->device_mutex);
+    return -EINVAL;
+  }
+
+  const loff_t new_file_position =
+      fixed_size_llseek(filp, offset, whence, (loff_t)circular_buffer_size);
+
+  // TODO: validate new_file_position
+  if (new_file_position < 0) {
+    PDEBUG("seek position (%lld) resulted in a negative value", offset);
+    mutex_unlock(&dev_ptr->device_mutex);
+    return -EINVAL;
+  }
+
+  if (new_file_position > circular_buffer_size) {
+    PDEBUG("seek position (%lld) exceeds circular buffer size", offset);
+    mutex_unlock(&dev_ptr->device_mutex);
+    return -EINVAL;
+  }
+
+  mutex_unlock(&dev_ptr->device_mutex);
```

Comment:
Based on in-class discussion, this unlock should be moved right after getting the `circular_buffer_size` instead of holding it the whole time.

Reply:
Yes